src/createUniqueTag.ts

```diff
@@ -0,0 +1,70 @@
+/**
+ * Global counter to scope the custom elements
+ *
+ * @type {number}
+ */
+let counter: number = Math.round(Math.random() * 100000);
+
+/**
+ * Allowed tag name chars
+ *
+ * @type {string}
+ */
+const chars: string = `-|\\.|[0-9]|[a-z]`;
+
+/**
+ * Regular expression to check if a value is a valid tag name
+ *
+ * @type {RegExp}
+ */
+const tagRegExp: RegExp = new RegExp(`[a-z](${chars})*-(${chars})*`);
+
+/**
+ * Checks if the tag name is valid
+ *
+ * @param {string} tag
+ * @returns {boolean}
+ */
+const isValid = (tag: string): boolean => tagRegExp.exec(tag) !== null;
+
+/**
+ * Checks if the tag is already registered
+ *
+ * @param {string} name
+ * @param {CustomElementRegistry} registry
+ * @returns {boolean}
+ */
+const isTagRegistered = (name: string, registry: CustomElementRegistry): boolean => !!registry.get(name);
+
+/**
+ * Given a tag name scopes it with a number suffix
+ *
+ * @param {string} tagName
+ * @param {CustomElementRegistry} registry
+ * @returns {string} scoped tag name
+ */
+const incrementTagName = (tagName: string, registry: CustomElementRegistry): string => {
+  const newTagName = `${tagName}-${(counter += 1)}`;
+
+  if (isTagRegistered(newTagName, registry)) {
+    return incrementTagName(tagName, registry);
+  }
+
+  return newTagName;
+};
+
+/**
+ * Creates a unique scoped tag name
+ *
+ * @exports
+ * @param {string} tagName - tag name to scope
+ * @param {CustomElementRegistry} registry
+ * @returns {string} scoped tag name
+ */
+export function createUniqueTag(tagName: string, registry: CustomElementRegistry = customElements): string {
```

Comment:
I don't think that classes passed through should have unique tags created for them. This would bloat the code running in the browser significantly. In this implementation, if a React dev using reactify were to write the following: 
```jsx
import { aButton } from "my-web-component-button";

const RedButton = reactify('a-button', { klass: aButton });
const BlueButton = reactify('a-button', { klass: aButton });
```
the browser would create two new CustomElements for the same code. It wouldn't extend either, these would be two new entire class definitions instead of instantiations of the same class.

Reply:
I ultimately think that if a web component klass is supplied with a name that already exists, the existing web component should be used. However, I'm not super satisfied with that implementation either, since that means we're ignoring times where a developer may be adding a class that's already defined... Any thoughts?

Reply:
I was doing this because of [scoped elements](https://open-wc.org/scoped-elements/). Some components libraries are exporting the class only to be able to scope the components. It's usefull when you do microfrontends because you can have multiple versions of components in the same page.

I could maybe remove the custom element definition here and keep only on my side.

Reply:
@MathieuPuech good catch that is how it works ! you let the user that composes the components handle the registration. You can add a version number to the component tag if you know they get often reused but that advanced my-comp-version

@BBKolton i have made a scopedElement registristrie impl that does validate the registered class so they get same class back on multi register

but conclusion is to handle registration on the integrator side not the component side,